File: src/main/java/com/faendir/acra/ui/view/base/navigation/Path.java
/*
 * (C) Copyright 2018 Lukas Morawietz (https://github.com/F43nd1r)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.faendir.acra.ui.view.base.navigation;

import com.faendir.acra.ui.navigation.MyNavigator;
import com.faendir.acra.ui.navigation.SingleViewProvider;
import com.vaadin.icons.VaadinIcons;
import com.vaadin.shared.ui.ContentMode;
import com.vaadin.ui.Button;
import com.vaadin.ui.Composite;
import com.vaadin.ui.CssLayout;
import com.vaadin.ui.Label;
import com.vaadin.ui.themes.AcraTheme;
import com.vaadin.ui.themes.ValoTheme;

import java.util.ArrayDeque;
import java.util.Deque;
import java.util.stream.Collectors;

/**
 * @author Lukas
 * @since 24.03.2018
 */
public class Path extends Composite {
    private final Deque<Element> elements;
    private final CssLayout layout;

    public Path() {
        elements = new ArrayDeque<>();
        layout = new CssLayout();
        layout.addStyleName(AcraTheme.BASIC_FLEX);
        setCompositionRoot(layout);
    }

    public void set(Deque<MyNavigator.HierarchyElement> hierarchy) {
        elements.clear();
        layout.removeAllComponents();
        hierarchy.forEach(e -> {
            SingleViewProvider provider = (SingleViewProvider) e.getProvider();
            goTo(provider.getTitle(provider.getParameters(e.getNavState())), e.getNavState());
        });
    }

    public void goTo(String label, String id) {
        goTo(new Element(label, id));
    }

    public void goTo(Element element) {
        if (elements.stream().map(Element::getId).anyMatch(element.getId()::equals)) {
            while (!getLast().getId().equals(element.getId())) {
                goUp();
            }
        } else {
            if (!elements.isEmpty()) {
                Label icon = new Label(VaadinIcons.CARET_RIGHT.getHtml(), ContentMode.HTML);
                layout.addComponent(icon);
            }
            Button button = new Button(element.getLabel());
            button.addClickListener(e -> {
                while (getLast() != element) goUp();
                getUI().getNavigator().navigateTo(asUrlFragment());
            });
            button.addStyleNames(ValoTheme.BUTTON_BORDERLESS, AcraTheme.PATH_ELEMENT);
            layout.addComponent(button);
            elements.addLast(element);
        }
    }

    public int getSize() {
        return elements.size();
    }

    public boolean isEmpty() {
        return elements.isEmpty();
    }

    public Element getLast() {
        return elements.getLast();
    }

    public Element goUp() {
        if (elements.isEmpty()) {
            return null;
        }
        removeLastComponent();
        if (elements.size() != 1) removeLastComponent();
        return elements.removeLast();
    }

    public void clear() {
        layout.removeAllComponents();
        elements.clear();
    }

    public String asUrlFragment() {
        return elements.stream().map(Element::getId).collect(Collectors.joining(MyNavigator.SEPARATOR));
    }

    private void removeLastComponent() {
        layout.removeComponent(layout.getComponent(layout.getComponentCount() - 1));
    }

    public static class Element {
        private final String label;
        private final String id;

        public Element(String label, String id) {
            this.label = label;
            this.id = id;
        }

        public String getLabel() {
            return label;
        }

        public String getId() {
            return id;
        }
    }
}

File: acrarium/src/main/kotlin/com/faendir/acra/ui/component/tabs/Path.kt
/*
 * (C) Copyright 2018 Lukas Morawietz (https://github.com/F43nd1r)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.faendir.acra.ui.component.tabs

import com.faendir.acra.i18n.TranslatableText
import com.faendir.acra.ui.component.Translatable
import com.vaadin.flow.component.AttachEvent
import com.vaadin.flow.component.Component
import com.vaadin.flow.component.DetachEvent
import com.vaadin.flow.component.tabs.Tab
import com.vaadin.flow.router.AfterNavigationEvent
import com.vaadin.flow.router.AfterNavigationListener
import com.vaadin.flow.shared.Registration
import org.springframework.context.support.GenericApplicationContext
import kotlin.reflect.KClass

/**
 * @author lukas
 * @since 18.10.18
 */
class Path(private val applicationContext: GenericApplicationContext) : SubTabs(), AfterNavigationListener {
    private var registration: Registration? = null

    override fun afterNavigation(event: AfterNavigationEvent) {
        removeAll()
        for (element in event.activeChain.filterIsInstance<HasRoute>().flatMap { it.getPathElements(applicationContext, event) }.reversed()) {
            val tab = element.toTab()
            add(tab)
            selectedTab = tab
        }
    }

    override fun onAttach(attachEvent: AttachEvent) {
        super.onAttach(attachEvent)
        registration = attachEvent.ui.addAfterNavigationListener(this)
    }

    override fun onDetach(detachEvent: DetachEvent) {
        registration?.remove()
        super.onDetach(detachEvent)
    }

    open class Element<T : Component>(val target: KClass<T>, val targetParams: Map<String, String>, titleId: String, vararg params: Any) :
        TranslatableText(titleId, *params) {
        fun toTab(): Tab {
            val link = Translatable.createRouterLink(target, targetParams, id, *params)
            link.addTranslatedListener { link.content.text?.let { link.content.element.setProperty("innerHTML", it.replace(".", ".<wbr>")) } }
            link.style["display"] = "block"
            link.style["overflow-wrap"] = "break-word"
            link.setWidthFull()
            return Tab(link)
        }
    }
}